File: app/src/main/java/com/app/cricapi/MainActivity.java
package com.app.cricapi;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.app.cricapi.adapter.MatchAdapter;
import com.app.cricapi.model.Match;
import com.app.cricapi.model.Root;
import com.app.cricapi.services.CricApiServices;
import com.google.gson.Gson;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class MainActivity extends AppCompatActivity {

    RecyclerView recyclerView;
    Button btnGetNewMatches;
    Retrofit retrofit;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        recyclerView=findViewById(R.id.recyclerView);
        btnGetNewMatches=findViewById(R.id.btnGetNewMatches);

        btnGetNewMatches.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Toast.makeText(MainActivity.this, "Getting data...", Toast.LENGTH_SHORT).show();

                retrofit=new Retrofit.Builder()
                        .baseUrl("https://cricapi.com/api/")
                        .addConverterFactory(GsonConverterFactory.create())
                        .build();

                CricApiServices apiServices=retrofit.create(CricApiServices.class);

                Call<Root> call=apiServices.getNewMatches();

                call.enqueue(new Callback<Root>() {
                    @Override
                    public void onResponse(Call<Root> call, Response<Root> response) {
                        if(response.isSuccessful()){
                            Root root=response.body();
                            List<Match> matchList=root.getMatches();

                            MatchAdapter matchAdapter=new MatchAdapter(MainActivity.this,matchList);
                            recyclerView.setLayoutManager(new LinearLayoutManager(MainActivity.this));
                            recyclerView.setAdapter(matchAdapter);
                        }
                        else {
                            Toast.makeText(MainActivity.this, response.code()+" "+response.message(), Toast.LENGTH_SHORT).show();
                        }
                    }

                    @Override
                    public void onFailure(Call<Root> call, Throwable t) {
                        Toast.makeText(MainActivity.this, "Something went wrong\n"+t.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });

            }
        });



    }
}

File: app/src/main/java/com/app/cricapi/MainActivity.kt
package com.app.cricapi

import android.os.Bundle
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.app.cricapi.adapter.MatchAdapter
import com.app.cricapi.model.Root
import com.app.cricapi.services.CricApiServices
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class MainActivity : AppCompatActivity() {

    var recyclerView: RecyclerView? = null
    var btnGetNewMatches: Button? = null
    var retrofit: Retrofit? = null

    protected override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        recyclerView = findViewById<RecyclerView>(R.id.recyclerView)
        btnGetNewMatches = findViewById<Button>(R.id.btnGetNewMatches)

        btnGetNewMatches!!.setOnClickListener {
            Toast.makeText(this@MainActivity, "Getting data...", Toast.LENGTH_SHORT).show()
            retrofit = Retrofit.Builder()
                    .baseUrl("https://cricapi.com/api/")
                    .addConverterFactory(GsonConverterFactory.create())
                    .build()
            val apiServices = Retrofit.Builder()
                    .baseUrl("https://cricapi.com/api/")
                    .addConverterFactory(GsonConverterFactory.create())
                    .build().create(CricApiServices::class.java)
            val call = apiServices.newMatches
            call.enqueue(object : Callback<Root?> {
                override fun onResponse(call: Call<Root?>, response: Response<Root?>) {
                    if (response.isSuccessful) {
                        val root = response.body()
                        val matchList = root!!.matches
                        val matchAdapter = MatchAdapter(this@MainActivity, matchList)
                        recyclerView!!.layoutManager = LinearLayoutManager(this@MainActivity)
                        recyclerView!!.adapter = matchAdapter
                    } else {
                        Toast.makeText(this@MainActivity, response.code().toString() + " " + response.message(), Toast.LENGTH_SHORT).show()
                    }
                }

                override fun onFailure(call: Call<Root?>, t: Throwable) {
                    Toast.makeText(this@MainActivity, """
     Something went wrong
     ${t.message}
     """.trimIndent(), Toast.LENGTH_SHORT).show()
                }
            })
        }
    }
}